File: benchmarks/goto_bench.py
"""Двойник ``goto_bench.bsl`` с тем же автоматом из восьми состояний.

В Python нет штатного ``goto``, поэтому переход к блоку представлен
ветвью ``if/elif``. Число проходов, порядок состояний и контрольная сумма
совпадают с BSL- и Lua-сценариями.
"""

from time import perf_counter


iteration = 0
state = 0
total = 0
started = perf_counter()

while iteration < 2_000_000:
    if state == 0:
        total += 1
        state = 1
    elif state == 1:
        total += 3
        state = 2
    elif state == 2:
        total += 5
        state = 3
    elif state == 3:
        total += 7
        state = 4
    elif state == 4:
        total += 11
        state = 5
    elif state == 5:
        total += 13
        state = 6
    elif state == 6:
        total += 17
        state = 7
    else:
        total += 19
        state = 0
    iteration += 1

elapsed_ms = (perf_counter() - started) * 1_000
print(f"сумма: {total}")
print(f"{elapsed_ms:.3f}")

File: benchmarks/goto_bench.lua
-- Двойник goto_bench.bsl: тот же граф из восьми блоков и те же
-- условные и безусловные переходы. Все локальные объявлены до первой
-- метки, поэтому Lua разрешает обратный переход к диспетчеру.

local iteration = 0
local state = 0
local sum = 0
local started = os.clock()

::dispatcher::
if state == 0 then goto branch0 end
if state == 1 then goto branch1 end
if state == 2 then goto branch2 end
if state == 3 then goto branch3 end
if state == 4 then goto branch4 end
if state == 5 then goto branch5 end
if state == 6 then goto branch6 end
goto branch7

::branch0::
sum = sum + 1
state = 1
goto next_iteration

::branch1::
sum = sum + 3
state = 2
goto next_iteration

::branch2::
sum = sum + 5
state = 3
goto next_iteration

::branch3::
sum = sum + 7
state = 4
goto next_iteration

::branch4::
sum = sum + 11
state = 5
goto next_iteration

::branch5::
sum = sum + 13
state = 6
goto next_iteration

::branch6::
sum = sum + 17
state = 7
goto next_iteration

::branch7::
sum = sum + 19
state = 0

::next_iteration::
iteration = iteration + 1
if iteration < 2000000 then goto dispatcher end

local elapsed_ms = (os.clock() - started) * 1000
print(string.format("сумма: %d", sum))
print(string.format("%.3f", elapsed_ms))
